samza-api/src/main/java/org/apache/samza/operators/functions/SinkFunction.java

```diff
@@ -42,6 +42,6 @@
    * @param messageCollector  the {@link MessageCollector} to send the message
    * @param taskCoordinator  the {@link TaskCoordinator} to request commits or shutdown
    */
-  void apply(M message, MessageCollector messageCollector, TaskCoordinator taskCoordinator);
+  M apply(M message, MessageCollector messageCollector, TaskCoordinator taskCoordinator);
```

Comment:
The sink _operator_ needs to return the original MessageStream it was applied on. The function doesn't need to return anything

Reply:
+1

It introduces confusion from the perspective of how is this even different from MapFunction?
Further, it opens up possibility for applications to return a different message than the input which makes it harder to reason about the semantics of SinkFunction.

Reply:
@bharathkk sink function is meath as an "escape hatch", in case something isn't currently expressible in the high level API. We should already treat it as a black-box. On second thought, might not be  bad idea to let it return a value, although in that case the new MessageStream should have the returned value, not original. WDYT?